Use native vectorization in Lucene

shubhamvishu opened the pull request:
### Description

This is continuation of the work @goankur started in https://github.com/apache/lucene/pull/13572. Quoting the [comment](https://github.com/apache/lucene/pull/13572#issuecomment-3648627594) below from the initial PR. 

> Here at Amazon (customer facing product search), we’ve been testing this native dot product implementation in our production environment(ARM - Graviton 2 and 3) and we see **5-14x** faster dot product computations in JMH benchmarks and we observed semantic latency improving from **62 msec** to **28 msec** (avg) for 4K embeddings(4.5 MM). Overall we saw **10-60%** improvement on end-end avg search latencies in different scenarios (different sized vectors, vector-focused search vs search combined with other workloads). We haven’t tested all other CPUs types yet. I'm working on a draft PR on top of this PR with following changes and planning to raise it soon :
> 
> * Removing the overhead from heap to off-heap copying by utilizing `Linker.Option.critical`, which eliminates unnecessary copying
> * Runtime dispatch using `IFUNC` to choose SVE vs NEON vs scalar implementation at runtime based on available intrinsics
> * Build related changes to generate the binary
> 
> We kept the native code isolated in the misc package and not getting it in the core module which we know is highly discouraged. Additionally, PR #15285 would later help eliminate some code duplication and enable a cleaner implementation similar to `PanamaVectorUtilSupport` - potentially through a `NativeVectorUtilSupport` class?
> 
> Our benchmarking suggests substantial optimization potential for ARM-based deployments, and we believe this could benefit the broader Lucene community. We hope to make it easy for any Lucene user to opt-in to this alternative vector implementation ideally. We're committed to refining this implementation based on community feedback and addressing any concerns during the review process. I'm eager to hear the community's thoughts on this change, as there appears to be significant optimization potential for ARM architectures that could benefit many users. Thank you!

To make use of faster SIMD instructions, `PanamaVectorUtilSupport#{dotProduct|uint8DotProduct}` will switch to native dot product implementation when `org.apache.lucene.util.Constants#NATIVE_DOT_PRODUCT_ENABLED` is `true` (i.e. system property`-Dlucene.useNativeDotProduct=true` is passed). 

GCC version : >= 10

<!--
If this is your first contribution to Lucene, please make sure you have reviewed the contribution guide.
https://github.com/apache/lucene/blob/main/CONTRIBUTING.md
-->

shubhamvishu committed:
Use native dot product in Lucene

shubhamvishu committed:
Fix ecjlint issues

shubhamvishu committed:
Make errorprone and changelog verifier happy

shubhamvishu committed:
compile it for windows

shubhamvishu committed:
Update library name for windows

shubhamvishu committed:
Update CHANGES.txt

shubhamvishu committed:
Typo and minor

dweiss commented on build.gradle:
```diff
@@ -18,6 +18,7 @@
 plugins {
   id "base"
   id "lucene.root-project.setup"
+  id "c"
```
this can be moved to the project which actually uses the plugin (misc?).

dweiss commented on lucene/core/build.gradle:
```diff
@@ -31,6 +31,14 @@ dependencies {
+          project(":lucene:misc").layout.buildDirectory.get().asFile.absolutePath + "/libs/dotProduct/shared"
+  )
+}
+
```
All of these changes should be pulled into a single gradle java plugin that handles the configuration across all the involved projects. If native libs are not enabled, these changes shouldn't be applied at all. A single plugin will make it easier to see the set of changes applied globally; currently they're scattered around.

dweiss reviewed (commented):
I'm not sure about this. The integration with native libs here is... tight. And to get a working version, you pretty much have to recompile from scratch on the target machine.

It would be more elegant to have it somehow wrapped in a service and a separate (optional) module. I'm not sure if it's possible.

Also - as I mentioned - some thought needs to be given to what the public artifacts are going to be - are binaries going to be shipped, for which cpus, how are they going to be compiled and what dev requirements this entails (installing cross-compilation environments is probably not going to fly with many).

dweiss commented:
For the record, any Java library that ships with native libs has similar integration problems. You can look at jansi or any other lib that has native components for inspiration (and to get an idea) how hairy it becomes. Here's jansi's docker-based cross-compilation build.

https://github.com/fusesource/jansi/blob/master/Makefile

shubhamvishu commented:
Thanks @dweiss for taking a look. I completely agree with maintaining different native bindings in Lucene can get really challenging as you mentioned and I propose we **don't ship don't ship any pre-built binaries** (or to say at least in this PR today if someone feels otherwise). I got some cross environment binaries working but I agree we might not be able to support all permutations/combinations of the different environments in the best possible way. 

While I got some cross-platform binaries working for few selective environments, supporting all environment permutations will be really painful. I added the build configuration to test across different environments on my end. I'm more happy to remove the Gradle build configurations for binary generation.

One important point here is that GCC's auto-vectorization of simple C code performs comparably to (and sometimes better than) hand-written NEON/SVE implementations(except some cases/runs for SVE). This allows us to keep things simple on Lucene's end and let users experiment. This is inline with what @rmuir proposed [in this comment](https://github.com/apache/lucene/pull/13572#issuecomment-2230061569). 

So here is what I propose we can do and would look like for user :

- We could keep the simple dot product implementation in the C code in `dotProduct.c` in the `lucene/misc` as-is but not build it. 
 
- Instead, we let the Lucene user to optionally generate the binary on their end using `gcc -shared -O3 -march=native -funroll-loops -o <c_file>` on the platform they wish to run (or) they would need to do so for whichever compiler and environment they are targeting(we can add more documentation if req.). User could either use the default `dotProduct.c` file with auto-vectorization or get creative and provide a more performant implementation if they wish to.

- To enable the native dot product codepath they could pass the `-Dlucene.useNativeDotProduct=true` system property.

**Contract:** This PR would enable the interface to interact with binaries for dot product computations but providing that binary is in user's court. Users need to provide the binary with the required signature/impl: 

`int32_t dot8s(int8_t vec1[], int8_t vec2[], int32_t limit);`

Let me know if this makes it clear for what it would look like from Lucene user's perspective.

dweiss commented:
I get the performance improvement is nice but it'll be difficult to maintain and test properly the way it's currently implemented. Not to mention most folks out there use Lucene without the possibility to recompile from sources....

I don't mind adding a native implementation but maybe this should be integrated via more standard mechanism (like service lookup) and then everything would live in that optional "native" module? This would open up the possibility to compile such a module independently. Sorry for my lack of enthusiasm about this but I can tell it'll be a problem to maintain even by looking at it (properties, etc.).

rmuir commented:
> For the record, any Java library that ships with native libs has similar integration problems. You can look at jansi or any other lib that has native components for inspiration (and to get an idea) how hairy it becomes. Here's jansi's docker-based cross-compilation build.
> 
> https://github.com/fusesource/jansi/blob/master/Makefile

i'd use zig

shubhamvishu committed:
Add pluggable(opt-in) NativeVectorSupport

shubhamvishu committed:
Revert gradle and VectorUtilBenchmark changes

shubhamvishu commented:
@dweiss I tried to make it more pluggable and added the `NativeVectorUtilSupport` which let users opt-in to use native binaries on the fly without doing anything extra. If there is library present and also have the required implementation it automatically switches to native implementation or else fallback to Panama based implementation. So nothing changes for users who don't want to use or try it(they would continue using `PanamaVectorUtilSupport` basically). I removed all the gradle related changes and not generate any binary at all for now. To to test a user could just simply :

```
// Passes

./gradlew test  
./gradlew test  -Ptest.native.dotProduct=false
./gradlew test -Ptest.native.dotProduct=true -Ptests.jvmargs="-Djava.library.path=/home/simd" 
./gradlew test -Ptests.jvmargs="-Dlucene.useNativeDotProduct=true -Djava.library.path=/home/simd"  


// Fails
./gradlew test -Ptests.jvmargs="-Dlucene.useNativeDotProduct=true"  
./gradlew test -Ptest.native.dotProduct=true
```

Note : I kept the C code (as is in misc) even though we are not building since it could be used as a reference for users or provides a good starting point.

&nbsp;

shubhamvishu committed:
Vector util service provider

shubhamvishu committed:
add service for native vector provider

shubhamvishu committed:
Some more changes and cleanup

dweiss commented:
Please give me some time - I've been away on a winter break.

shubhamvishu committed:
Merge branch 'main' into simd

shubhamvishu committed:
impl new int4DibitDotProduct

shubhamvishu committed:
Allows providing native impl for other VectorUtilSupport funcs

shubhamvishu committed:
fix errorprone

shubhamvishu commented:
Hi @dweiss , I made some more changes here and looking forward to your feedback if you get chance. Thank you!

------
Unrelated: GitHub Actions appears to be experiencing an outage - workflows are stuck in the queue. https://www.githubstatus.com/
> Update - The team continues to investigate issues causing GitHub Actions jobs on hosted runners to remain queued for extended periods, with a percentage of jobs failing. We will continue to provide updates as we make progress toward mitigation.
 Feb 02, 2026 - 20:27 UTC

shubhamvishu committed:
Log info in NativeVectorizationProvider

shubhamvishu commented:
FYI we now have a bug/issue filed in OpenJDK to improve vector operation support on AArch64 machines : https://bugs.openjdk.org/browse/JDK-8377386

shubhamvishu committed:
Merge branch 'main' into simd

shubhamvishu committed:
Revert Vector util service provider impl

dweiss commented:
Yes, I think so.

shubhamvishu committed:
Minor changes

shubhamvishu commented:
@dweiss I made those changes in the latest revision. Thank you!

shubhamvishu committed:
Refactor to make less intrusive

shubhamvishu committed:
cleanup

shubhamvishu committed:
Merge branch 'main' into simd

shubhamvishu commented:
@dweiss I pushed a version that's much less intrusive and tried to simplify/refactor some bits as you suggested. Now we do not even initialize the `NativeVectorUtilSupport` class(confirmed by deliberately throwing an exception in static block) unless If the user has user has opted-in for it by passing `-Dlucene.useNativeDotProduct=true` property. So the normal/current codepath remains unchanged and continues with Panama.

 Let me know your thoughts!

shubhamvishu committed:
Merge branch 'main' into simd

shubhamvishu committed:
fix some recent changes to CHANGES.txt

shubhamvishu commented:
Thank you @dweiss ! I'll wait for a few days to see if anyone has more feedback and then merge.

shubhamvishu commented on lucene/core/src/java25/org/apache/lucene/internal/vectorization/NativeVectorUtilSupport.java:
```diff
@@ -0,0 +1,590 @@
+  private static <T> T invokeOrDelegate(MethodHandle mh, Supplier<T> delegate, Object... args) {
+    if (mh != null) {
+      try {
+        return (T) mh.invokeExact(args);
```
I opened https://github.com/apache/lucene/issues/15840 to discuss the ideas on improving the test coverage for the native path in our CI workflows.

uschindler reviewed (commented):
Please update this to make it consistent. No class except the official entry points in the internal vectorization package should be public
